api/C/include/csi_ll_pmp.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Low-Level PMP API
+ *
+ * This module provides functionality for PMP configuration.
+ *
+ * Copyright (C) 2017-2023 Alibaba Group Holding Limited
+ */
+
+#ifndef CSI_LL_PMP_H
+#define CSI_LL_PMP_H
+
+#include "csi_types.h"
+#include <stdint.h>
+
+/*
+ * Types of address Matching
+ */
+typedef enum {
+    ADDRESS_MATCHING_OFF   = 0,	/* Null region (disabled) */
+    ADDRESS_MATCHING_TOR   = 1, /* Top of range */
+    ADDRESS_MATCHING_NA4   = 2, /* Naturally aligned four-byte region */
+    ADDRESS_MATCHING_NAPOT = 3, /* Naturally aligned power-of-two region, ≥8 bytes */
+} csi_addr_matching_t;
+
+/*
+ * Attribution of PMP coniguration
+ */
+typedef struct {
+    bool flag_r; /* The flag of readable. 0: not readable; 1: readable */
+    bool flag_w; /* The flag of writeable. 0: not writeable; 1: writeable */
+    bool flag_x; /* The flag of execable. 0: not execable; 1: execable */
+    csi_addr_matching_t mode; /* Address matching mode */
+    bool lock; /* The flag of lock,  0: not be locked; 1 : be locked */
+} csi_mpu_region_attr_t;
+
+/*
+ * Configure PMP range.
+ *
+ * This function only can be called in M-mode. If it is called
+ * in other mode, it may return an error or cause carsh. When this function is called
+ * with success, it will not be effective untill the core has been exit M-mode.
+ *
+ * @param idx: The index of PMP config entries
+ * @param base_addr: The base address of this range need to be configured
+ * @param size: The size of this range need to be configured
+ * @param attr: A point to the attribution of this range to be configured
+ * @param enable: Eanble flag to indicate whether this ranged can be accessed or not
+ * @return : Status code
+ */
+csi_status_t csi_mpu_config_region(unsigned int idx, unsigned long base_addr,
```

Comment:
This API is OK for applications with a fixed PMP setup.  But it is slow, so maybe not suitable for systems where the PMP setup changes when context-switching between tasks.  Perhaps there should be another function with similar parameters, which can be used to construct a PMP setup structure in memory; so the data in this structure can then be written quickly into the CSRs?

Reply:
For such application in MCU, it is better to realize functions by assemble code instead of LL code.  For PMP context size, it is related with CPU PMP configuration.  I strongly recommend PMP entry size is 16 at first, it is popular configuration in practise else.

Reply:
I don't agree that assembly code is necessary to deal with this, it can all be more easily and efficiently handled from C code.  You can allow the BSP to control context size - simply let the BSP define a context structure which the user application instantiates and passes in to the API for initialization.  But most uses of the PMP require dynamic switching of PMP setup according to what process is running.  This means constructing the register values in advance, to make context switch more efficient.
If we add checking for invalid setup into the definition of what this function does, then this makes it even more inefficient to execute on a context switch.

Reply:
To be clear, I am suggesting 2 additional C APIs: one with a similar parameter set to this one, which can be used to construct a PMP setup within a context structure.  And another which quickly and efficiently writes the contents of that context structure into the CSRs.

Reply:
I agree with @WiseCrohn to provide some lower level pmp api and a little higher level api for flexiable configuration of pmp region configuration.

In NMSIS, we defined it like these 

1. PMP API Header file: https://github.com/Nuclei-Software/NMSIS/blob/master/NMSIS/Core/Include/core_feature_pmp.h#L58-L379
2. PMP API Doc: https://doc.nucleisys.com/nmsis/core/api/core_pmp.html
3. PMP Example: https://github.com/Nuclei-Software/nuclei-sdk/blob/master/application/baremetal/demo_pmp/demo_pmp.c#L53-L129